test/test_decorator.py

```diff
@@ -0,0 +1,117 @@
+"""Tests simple parsing decorators.
+"""
+import collections
+import dataclasses
+import functools
+import inspect
+import typing
+from typing import Callable
+
+import pytest
+import simple_parsing as sp
+
+
+@dataclasses.dataclass
+class AddThreeNumbers:
+    a: int = dataclasses.field(default=0)
+    b: int = dataclasses.field(default=0)
+    c: int = dataclasses.field(default=0)
+
+    def __call__(self) -> int:
+        return self.a + self.b + self.c
+
+
+def _fn_with_positional_only(x: int, /) -> int:
+    return x
+
+
+def _fn_with_keyword_only(*, x: int) -> int:
+    return x
+
+
+def _fn_with_all_argument_types(a: int, /, b: int, *, c: int) -> int:
+    return a + b + c
+
+
+class ListWithPopDefault(collections.UserList):
+    def pop(self, index, default_value=None):
+        try:
+            return super().pop(index)
+        except IndexError:
+            return default_value
+
+
+def partial(fn: Callable, *args, **kwargs) -> Callable:
+    """Partial via changing the signature defaults."""
+
+    @functools.wraps(fn)
+    def _wrapper(*other_args, **other_kwargs):
+        return fn(*other_args, **other_kwargs)
+
+    signature = inspect.signature(fn)
+    parameters = signature.parameters
+
+    args = ListWithPopDefault(args)
+
+    new_parameters = []
+    for key, param in parameters.items():
+        param = typing.cast(inspect.Parameter, param)
+        default = args.pop(0, None) or kwargs.get(key, None)
+        if default is not None:
+            param = param.replace(default=default)
+        new_parameters.append(param)
+
+    signature = signature.replace(parameters=new_parameters)
+    _wrapper.__signature__ = signature
+
+    return _wrapper
+
+
+@pytest.mark.parametrize(
+    "args, expected, fn",
+    [
+        ("", 1, partial(_fn_with_positional_only, 1)),
+        ("2", 2, partial(_fn_with_positional_only, 1)),
+        ("2", 2, _fn_with_positional_only),
+        ("", 1, partial(_fn_with_keyword_only, x=1)),
+        ("--x=2", 2, partial(_fn_with_keyword_only, x=1)),
+        ("--x=2", 2, _fn_with_keyword_only),
+        ("", 3, partial(_fn_with_all_argument_types, 1, b=1, c=1)),
+        ("2", 4, partial(_fn_with_all_argument_types, b=1, c=1)),
+        ("2 --b=2", 5, partial(_fn_with_all_argument_types, c=1)),
+        ("2 --b=2 --c=2", 6, _fn_with_all_argument_types),
+        ("--c=2", 4, partial(_fn_with_all_argument_types, 1, b=1)),
+        ("--b=2", 4, partial(_fn_with_all_argument_types, 1, c=1)),
+        ("--b=2 --c=2", 5, partial(_fn_with_all_argument_types, 1)),
+    ],
+)
+def test_simple_arguments(
```

Comment:
These tests are fine for now, but in the future we'd need to make sure this works with other types of fields besides ints.

Just for fun, here's how I parameterize the field types for other tests, which I think is both very hacky and very pretty:

https://github.com/lebrice/SimpleParsing/blob/master/test/test_base.py#L81

Perhaps you could do something similar when constructing your functions (or I could add it myself after this PR).
Note: If I recall correctly, this kind of trickery only works when you type annotations are evaluated at runtime (as you are doing here).

Reply:
The reason I hadn't tested other fields is that those tests should already be covered by SP. We need to check that the decorator constructs the dataclass properly and everything else is handled by `sp.parse`. If we have test coverage for those types in `sp.parse` we should be good.

Reply:
Fair enough.